content/derived-datatypes.rst

```diff
@@ -11,47 +11,461 @@ Derived datatypes
 
+.. math::
+   :label: eq:typesig
 
+   \textrm{Typemap}[\texttt{T}] = [ \texttt{Datatype}_{0}, \ldots, \texttt{Datatype}_{n-1} ]
 
-.. code-block:: c
+The **typemap** is the associative array (map) with datatypes, as understood by MPI, as
+*keys* and displacements, in bytes, as *values*.
 
-   struct simple {
-    int an_integer;
-    char a_char;
-   };
 
-.. figure:: img/sample-image.png
-   :class: with-border
+.. math::
+   :label: eq:typemap
 
-   The relation between size and extent of a derived data type for the case of the ``simple`` datatype.
+   \textrm{Typemap}[\texttt{T}] = \{ \texttt{Datatype}_{0}: \textrm{Displacement}_{0}, \ldots, \texttt{Datatype}_{n-1}: \textrm{Displacement}_{n-1} \}
+
+
+The displacements are *relative* to the buffer the datatype describes.
+
+Assuming that an ``int`` takes 4 bytes of memory, the typemap for our ``Pair``
+datatype would be: :math:`\textrm{Typemap}[\texttt{Pair}] = \{ \texttt{int}: 0,
+\texttt{char}: 4\}`. Note again that the displacements are *relative*.
+
+.. figure:: img/E01-displacements.svg
+   :align: center
+
+   Depiction of the typemap for the ``Pair`` custom type. The displacements are
+   always relative.
+
+
+Knowledge of typemap and type signature is not enough for a full description of
+the type to the MPI runtime: the underlying programming language might mandate
+architecture-specific :term:`alignment` of the basic datatypes. The data
+structure would then be laid out in memory incoherently with the displacements
+in its typemap.
+We need a few more concepts. Given a typemap :math:`m` we can define:
+
+Lower bound
+  The first byte occupied by the datatype.
+
+  .. math::
+     :label: eq:lowerbound
+
+     \textrm{LB}[m] = \min_{j}[\textrm{Displacement}_{j}]
+
+Upper bound
+  The last byte occupied by the datatype.
+
+  .. math::
+     :label: eq:upperbound
+
+     \textrm{UB}[m] = \max_{j}[\textrm{Displacement}_{j} + \texttt{sizeof}(\textrm{Datatype}_{j})] + \textrm{Padding}
+
+Extent
+  The amount of memory needed to represent the datatype, taking into account architecture-specific alignment.
+
+  .. math::
+     :label: eq:extent
+
+     \textrm{Extent}[m] = \textrm{UB}[m] - \textrm{LB}[m]
+
+
+The C language (and Fortran) *require* that the data occurs in memory at
+well-defined addresses: the data needs to be aligned. The address, in bytes, of
+any item must be a multiple of the size of that item in bytes. This is so-called
+*natural alignment*.
+For our ``Pair`` data structure the ``first`` element is an ``int`` and occupies
+4 bytes. An ``int`` will align to 4 bytes boundaries: when allocating a new
+``int`` in memory, the compiler will insert **padding** to reach the alignment
+boundary.
+Indeed, ``second`` is a ``char`` and requires just 1 byte. To insert yet another
+``Pair`` item, we first need to reach the alignment boundary with a padding of 3
+bytes.
+Thus:
+
+.. math::
+
+   \begin{aligned}
```

Comment:
To be even more clear, you could enumerate the values of Displacement_j and sizeof(Datatype_j), so that the different values of 4 on line 178 are clearly understandable

Reply:
Good idea.